codegen/csg.adoc

```diff
@@ -8,4 +8,193 @@ permalink: /codegen_custom_swagger_gentemplate/
+executable jar file and the class that implements the selected SCG module as command-line
+arguments.footnote[Although SCG is published with several convenience scripts that invoke the CLI on
+specific sample models embedded in the project, it does not provide a general-purpose script to
+invoke the CLI. Such a script would be simple to create, but in this guide we will show the full
+Java command lines.]
+
+Our first step invokes the SCG `meta` command to create a skeleton project. The command arguments we
+will use are:
+
+`-o dir`:: Specifies the directory where the skeleton project will be created; default is current directory.
+
+`-n name`:: Specifies the name for the new SCG module, which is put to many uses in the skeleton,
+some of which imply restrictions on the name that are not enforced by SCG. These
+include:
+
+** It is the advertised name of the module (i.e. the value its `CodegenConfig#getName()` method
+   returns). The CLI `generate` command's `-l` (`--lang`) option maps to these names.
+** It is incorporated into the name of the generator class, which is the name value with its first
+   character upper-cased, concatenated with `Generator`. E.g. `-n myModule` would generate a class
+   named `MyModuleGenerator`.
+** It is incorporated into the advertised description of the model (the value returned by
+   `CodegenConfig#getHelp()).
+** It is incorporated into the name of the default output folder used by the module. (This is
+   unused when the module is executed in {RAS} as a GenTemplate, since the output directory is in
+   that case specified in the GenTarget.)
+** It is used as the template directory name. This directory will appear as a top-level Java
+   resource directory in the generated project.
+
+`-p package`:: Specifies the package in which the generator class will reside. The default is
+`io.swagger.codegen`.
+
+In preparation for this step we need to install the Swagger Codegen Client jar file. One way to do
+that is to clone the link:https://github.com/swagger-api/swagger-codegen[swagger-api/swagger-codegen
+git repo], checking out version 2.1.4 (using the git tag `v2.1.4`) and then running `mvn package`
+from the working tree root. This will build the needed jar file in
+[.nowrap]`modules/swagger-codegen-cli/target/swagger-codegen-cli.jar`.
+
+We'll execute this step by starting a command shell for our system and navigating to a directory
+that will hold the skeleton as a subdirectory. For simplicity we'll assume that parent directory is
+the user's home directory, designated in the linux style as `~`, and that the swagger-codegen git
+clone is located in `~/git/swagger-codegen`.
+
+After all the foregoing, this step finally comes down to a single command, which will create our
+skeleton:
+
+[source%nowrap]
+--
+java -jar ~/git/swagger-codegen/modules/swagger-codegen-cli/target/swagger-codegen-cli.jar meta -o ~/LexiconSCGentemplates -n lexiconSCGenTemplate -p my.org
```

Comment:
@tedepstein Sorry to say that almost everything you said here is not true! :)

First, the project directory need not be - and is not in the steps I've outlined - within the workspace. Eclipse (and therefore API Studio) supports "linked" projects that live outside the workspace root. In fact, these days there are probably more linked projects in the world than otherwise, since so many projects are imported from git clones that are created outside the workspace.

It's also not true that the scg-cli project lives - or needs to live - within the workspace, or that there are any classpath issues that arise with it. The jar file that's produced in that project is an "executable" jar file, which means it defines its own class path as well as an "entry point" class. When you run `java -jar xxx.jar`, Java loads that jar file, sets the classpath (ignoring any other ways of specifying classpath like environment variable or `-cp` options), and attempts to run the entry point class. So it doesn't matter where this jar file ends up, you just have to identify it correctly with the `-jar` option for it to work.

The `-o` option here really is a mouthful, but it also represents the most likely location for anyone following along with this. I expect that anyone attempting to create their own custom SCG module will completely understand what's going on here and simplify if they care to.

The real culprit here, IMO, is the swagger-codegen folks, who ship SCG with a couple dozen shell scripts that each make it easy to run a specific module against the petstore example and stick it in a specific output folder - all hard-wired. Yet they couldn't be bothered to provide a very simple but more general script that hides that big, ugly jar file path and leaves all the options open to the user!